src/xml.ts

```diff
@@ -0,0 +1,52 @@
+import { XmlReplacements } from "./types";
+import { notDeepEqual } from "assert";
+
+// Use a node name that someone else won't pick accidentally
+const XML_WRAPPER = "xml_wrapper_faeqcd"
+let parser: DOMParser;
+
+export default function parseXml<T>(xmlString: string, replacements: XmlReplacements<T>): (string | T)[] {
+	if (typeof parser === 'undefined') {
+		parser = new DOMParser();
+	}
+
+	const xmlDoc = parser.parseFromString(
+		`<${XML_WRAPPER}>${xmlString}</${XML_WRAPPER}>`,
```

Comment:
Why is this better than wrapping in a span and why does it matter if the name has been used before?

Reply:
I restructured the recursion to call `walk` on the root DOM node so that the `walk` function could own the logic about how to deal with node children. If we stuck with `span`, and `span` was used elsewhere in the XML string, the same replacement function used for the user's `span` would be used for the wrapper `span`, which we don't want. We want a custom wrapper that won't happen to match one of the replacements.

Reply:
This seems like an unacceptable security risk; there should never be a risk of having a magic string cause an extra component to end up in your array. Is there any way to avoid it?

Reply:
1. Security risk? Not sure how it has anything to do with security. No one should be passing user input to this function.
2. The risk is almost zero. Did you see the string? It's `xml_wrapper_faeqcd`. I can make the random part longer if you're truly concerned about a collision.

If you're really concerned I can look at restructuring again so the wrapper tag name doesn't matter, but I do like the elegance of the current function structure.

Reply:
Parsing XML calls for extra caution at the best of times. We're going a step further and asking the caller to trust that this library will take a dictionary of components and compose them safely and predictably. If one of those dictionary keys is a magic value which can cause a component to be rendered twice, once as the parent of the other components, that isn't predictable or safe. Or elegant, for that matter.

Reply:
I was careful not to use the word "elegant" to refer to the wrapper tag name, only the function structure it afforded. :)

I didn't really like the magic tag name that much, but I did like the way it let me restructure the functions to keep the node logic in one spot, and I didn't want to revert that. I changed things up though so that the wrapper tag name doesn't matter, but the function structure stays the same. Also started ignoring XML comments in the process. And added a couple more tests.